File: app/src/main/java/br/com/clust3rr/calculatordownload/MainActivity.java
package br.com.clust3rr.calculatordownload;

import android.content.DialogInterface;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.SeekBar;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {

    private static final double SIZE_BIT = 1;
    private EditText mFileSize, mConnectionSpeed;
    private Button mBtnFile, mBtnConnection;
    private TextView mResult, mTax;
    private static final int DEFAULT_MARGIN = 5;
    private SeekBar mMarginRate;
    private static final double SIZE_BYTE = SIZE_BIT * 8;
    private static final double SIZE_KB = 1000d * SIZE_BYTE;
    private static final double SIZE_MB = 1000 * SIZE_KB;
    private static final double SIZE_GB = 1000 * SIZE_MB;
    private static final double SIZE_TB = 1000 * SIZE_GB;
    private static final double SIZE_Kb = 1000 * SIZE_BIT;
    private static final double SIZE_Mb = 1000 * SIZE_Kb;
    private static final double SIZE_Gb = 1000 * SIZE_Mb;
    private static final double SIZE_Tb = 1000 * SIZE_Gb;
    private static final double SIZE_KiB = 1024 * SIZE_BIT;
    private static final double SIZE_MiB = 1024 * SIZE_KiB;
    private static final double SIZE_GiB = 1024 * SIZE_MiB;
    private static final double SIZE_TiB = 1024 * SIZE_GiB;
    private static final double VEL_bit_S = SIZE_BIT;
    private static final double VEL_byte_S = SIZE_BYTE;
    private static final double VEL_Kb_S = SIZE_Kb;
    private static final double VEL_KB_S = SIZE_KB;
    private static final double VEL_Mb_S = SIZE_Mb;
    private static final double VEL_MB_S = SIZE_MB;
    private static final double VEL_Gb_S = SIZE_Gb;
    private static final double VEL_GB_S = SIZE_GB;
    private static final double TIME_S = 1;
    private static final double TIME_MINUTE = TIME_S * 60;
    private static final double TIME_HOUR = TIME_MINUTE * 60;
    private static final double TIME_DAY = TIME_HOUR * 24;
    private static final double TIME_YEAR = TIME_DAY * 365;
    private String[] mCharSequenceFileType, mCharSequenceConnectionType;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mFileSize = findViewById(R.id.textInputEditFileSize);
        mConnectionSpeed = findViewById(R.id.textInputEditConnectionSpeed);
        mBtnFile = findViewById(R.id.buttonFile);
        mBtnConnection = findViewById(R.id.buttonConection);
        mMarginRate = findViewById(R.id.seekBar);
        mResult = findViewById(R.id.result);
        mTax = findViewById(R.id.error_margin);

        mMarginRate.setProgress(DEFAULT_MARGIN);
        mTax.setText(getString(R.string.errorMargin, DEFAULT_MARGIN));

        //
        mCharSequenceFileType = new String[]{
                getString(R.string.size_bits),
                getString(R.string.size_bytes),
                getString(R.string.size_KB),
                getString(R.string.size_MB),
                getString(R.string.size_GB),
                getString(R.string.size_TB),
                getString(R.string.size_Kb),
                getString(R.string.size_Mb),
                getString(R.string.size_Gb),
                getString(R.string.size_Tb),
                getString(R.string.size_KiB),
                getString(R.string.size_MiB),
                getString(R.string.size_GiB),
                getString(R.string.size_TiB)
        };

        mCharSequenceConnectionType = new String[]{
                getString(R.string.vel_bit_s),
                getString(R.string.vel_byte_s),
                getString(R.string.vel_Kb_s),
                getString(R.string.vel_KB_s),
                getString(R.string.vel_Mb_s),
                getString(R.string.vel_MB_s),
                getString(R.string.vel_Gb_s),
                getString(R.string.vel_GB_s)
        };

        mFileSize.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                updateResult();
            }
        });

        mConnectionSpeed.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                updateResult();
            }
        });

        mMarginRate.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                mTax.setText(getString(R.string.errorMargin, progress));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                updateResult();
            }
        });

        updateResult();
    }

    private void updateResult() {
        if (mFileSize != null && mConnectionSpeed != null && mResult != null) {
            String fs = mFileSize.getText().toString();
            String cs = mConnectionSpeed.getText().toString();
            int mr = mMarginRate.getProgress();

            if (fs.equals("") || cs.equals(""))
                mResult.setText(getString(R.string.noTime));
            else
                mResult.setText(calculateTimeToDownload(fs, cs, mr));

        }

    }

    public void buttonFileOnClick(View v) {
        AlertDialog.Builder dialog = new AlertDialog.Builder(this);
        dialog.setTitle(R.string.fileSize);
        int position = 0;
        for (int i = 0; !(mCharSequenceFileType[i].equals(mBtnFile.getText().toString())); i++)
            position = i + 1;

        dialog.setSingleChoiceItems(mCharSequenceFileType, position, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                mBtnFile.setText(mCharSequenceFileType[which]);
                updateResult();
                dialog.cancel();
            }
        }).create().show();
    }

    public void buttonConnectionOnClick(View v) {
        AlertDialog.Builder dialog = new AlertDialog.Builder(this);
        dialog.setTitle(R.string.connectionSpeed);
        int position = 0;
        for (int i = 0; !(mCharSequenceConnectionType[i].equals(mBtnConnection.getText().toString())); i++)
            position = i + 1;

        dialog.setSingleChoiceItems(mCharSequenceConnectionType, position, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                mBtnConnection.setText(mCharSequenceConnectionType[which]);
                updateResult();
                dialog.cancel();
            }
        }).create().show();
    }

    private String calculateTimeToDownload(String size, String speed, int margin) {

        double fileSize = 0, connectionSpeed = 0;

        try {
            fileSize = Double.parseDouble(size);
            connectionSpeed = Double.parseDouble(speed);
        } catch (NumberFormatException e) {
            e.printStackTrace();
        }

        double fileSize_bit = 0, connectionSpeed_bit = 0, timeInSeconds;

        String fileText = mBtnFile.getText().toString();

        if (mCharSequenceFileType[0].equals(fileText))
            fileSize_bit = fileSize * SIZE_BIT;
        else if (mCharSequenceFileType[1].equals(fileText))
            fileSize_bit = fileSize * SIZE_BYTE;
        else if (mCharSequenceFileType[2].equals(fileText))
            fileSize_bit = fileSize * SIZE_KB;
        else if (mCharSequenceFileType[3].equals(fileText))
            fileSize_bit = fileSize * SIZE_MB;
        else if (mCharSequenceFileType[4].equals(fileText))
            fileSize_bit = fileSize * SIZE_GB;
        else if (mCharSequenceFileType[5].equals(fileText))
            fileSize_bit = fileSize * SIZE_TB;
        else if (mCharSequenceFileType[6].equals(fileText))
            fileSize_bit = fileSize * SIZE_Kb;
        else if (mCharSequenceFileType[7].equals(fileText))
            fileSize_bit = fileSize * SIZE_Mb;
        else if (mCharSequenceFileType[8].equals(fileText))
            fileSize_bit = fileSize * SIZE_Gb;
        else if (mCharSequenceFileType[9].equals(fileText))
            fileSize_bit = fileSize * SIZE_Tb;
        else if (mCharSequenceFileType[10].equals(fileText))
            fileSize_bit = fileSize * SIZE_KiB;
        else if (mCharSequenceFileType[11].equals(fileText))
            fileSize_bit = fileSize * SIZE_MiB;
        else if (mCharSequenceFileType[12].equals(fileText))
            fileSize_bit = fileSize * SIZE_GiB;
        else if (mCharSequenceFileType[13].equals(fileText))
            fileSize_bit = fileSize * SIZE_TiB;

        String connectionText = mBtnConnection.getText().toString();

        if (mCharSequenceConnectionType[0].equals(connectionText))
            connectionSpeed_bit = connectionSpeed * VEL_bit_S;
        else if (mCharSequenceConnectionType[2].equals(connectionText))
            connectionSpeed_bit = connectionSpeed * VEL_byte_S;
        else if (mCharSequenceConnectionType[3].equals(connectionText))
            connectionSpeed_bit = connectionSpeed * VEL_Kb_S;
        else if (mCharSequenceConnectionType[4].equals(connectionText))
            connectionSpeed_bit = connectionSpeed * VEL_KB_S;
        else if (mCharSequenceConnectionType[5].equals(connectionText))
            connectionSpeed_bit = connectionSpeed * VEL_Mb_S;
        else if (mCharSequenceConnectionType[6].equals(connectionText))
            connectionSpeed_bit = connectionSpeed * VEL_MB_S;
        else if (mCharSequenceConnectionType[7].equals(connectionText))
            connectionSpeed_bit = connectionSpeed * VEL_Gb_S;
        else if (mCharSequenceConnectionType[8].equals(connectionText))
            connectionSpeed_bit = connectionSpeed * VEL_GB_S;

        timeInSeconds = fileSize_bit / (connectionSpeed_bit * (1 - ((double) margin / 100)));
        return timeToSting(timeInSeconds);
    }

    private String timeToSting(Double timeInSeconds) {

        double minutes = (timeInSeconds / TIME_MINUTE);
        double hours = (timeInSeconds / TIME_HOUR);
        double days = (timeInSeconds / TIME_DAY);

        String day = getString(R.string.day);
        String hour = getString(R.string.hour);
        String minute = getString(R.string.minute);
        String second = getString(R.string.second);

        if (timeInSeconds > TIME_YEAR) {
            return getString(R.string.highTime);
        } else if (timeInSeconds < 1) {
            return getString(R.string.lowTime);
        } else if (days >= 1) {
            return (int) Math.floor(days) + "" + day + " " +
                    (int) Math.floor(days * 24 % 24) + "" + hour + " " +
                    (int) Math.floor(days * 24 * 60 % 60) + "" + minute + " " +
                    (int) Math.floor(days * 24 * 60 * 60 % 60) + "" + second;
        } else if (hours >= 1) {
            return (int) Math.floor(hours) + "" + hour + " " +
                    (int) Math.floor(hours * 60 % 60) + "" + minute + " " +
                    (int) Math.floor(hours * 60 * 60 % 60) + "" + second;
        } else if (minutes >= 1) {
            return (int) Math.floor(minutes) + "" + minute + " " +
                    (int) Math.floor(minutes * 60 % 60) + "" + second;
        } else if (timeInSeconds >= TIME_S) {
            return (int) Math.floor(TIME_S) + "" + second;
        } else return getString(R.string.noTime);

    }
}

File: app/src/main/java/br/com/clust3rr/calculatordownload/MainActivity.kt
package br.com.clust3rr.calculatordownload

import android.content.DialogInterface
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.widget.SeekBar
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import br.com.clust3rr.calculatordownload.databinding.ActivityMainBinding
import java.lang.Math.pow
import kotlin.math.exp
import kotlin.math.floor

class MainActivity : AppCompatActivity(), TextWatcher, SeekBar.OnSeekBarChangeListener {
    private lateinit var listFileSizeMeasurementType: Array<String>
    private lateinit var listConnectionSpeedMeasurementType: Array<String>

    private val binding by lazy { ActivityMainBinding.inflate(layoutInflater) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)

        listFileSizeMeasurementType = arrayOf(
            getString(R.string.size_si_byte),

            getString(R.string.size_si_kilo_byte),
            getString(R.string.size_si_mega_byte),
            getString(R.string.size_si_giga_byte),
            getString(R.string.size_si_tera_byte),

            getString(R.string.size_iec_kilo_byte),
            getString(R.string.size_iec_mega_byte),
            getString(R.string.size_iec_giga_byte),
            getString(R.string.size_iec_tera_byte)
        )

        listConnectionSpeedMeasurementType = arrayOf(
            getString(R.string.speed_bit_s),
            getString(R.string.speed_kilo_bit_s),
            getString(R.string.speed_mega_bit_s),
            getString(R.string.speed_giga_bit_s),

            getString(R.string.speed_byte_s),
            getString(R.string.speed_kilo_byte_s),
            getString(R.string.speed_mega_byte_s),
            getString(R.string.speed_giga_byte_s)
        )

        binding.marginErrorSlider.progress = DEFAULT_ERROR_MARGIN
        binding.errorMarginDisplay.text = getString(R.string.errorMargin, DEFAULT_ERROR_MARGIN)

        binding.inputFileSize.addTextChangedListener(this)
        binding.inputConnectionSpeed.addTextChangedListener(this)
        binding.marginErrorSlider.setOnSeekBarChangeListener(this)

        binding.buttonFileSizeList.setOnClickListener { buttonFileOnClick() }
        binding.buttonConnectionSpeedList.setOnClickListener { buttonConnectionOnClick() }

        // setup default selection for measurement types
        binding.buttonFileSizeList.text = getString(R.string.size_si_giga_byte)
        binding.buttonConnectionSpeedList.text = getString(R.string.speed_mega_byte_s)

        updateResult()
    }

    private fun updateResult() {
        val fs = binding.inputFileSize.text.toString()
        val cs = binding.inputConnectionSpeed.text.toString()
        val mr = binding.marginErrorSlider.progress

        binding.timeResult.text = if(fs.isBlank() || cs.isBlank())
            getString(R.string.noTime)
        else
            calculateTimeToDownload(fs, cs, mr) ?: "Error"
    }

    private fun buttonFileOnClick() {
        val dialog = AlertDialog.Builder(this)
        dialog.setTitle(R.string.fileSize)
        var position = 0
        var i = 0
        while (listFileSizeMeasurementType[i] != binding.buttonFileSizeList.text.toString()) {
            position = i + 1
            i++
        }

        dialog.setSingleChoiceItems(
            listFileSizeMeasurementType,
            position
        ) { mDialog: DialogInterface, which: Int ->
            binding.buttonFileSizeList.text = listFileSizeMeasurementType[which]
            updateResult()
            mDialog.cancel()
        }.create().show()
    }

    private fun buttonConnectionOnClick() {
        val dialog = AlertDialog.Builder(this)
        dialog.setTitle(R.string.connectionSpeed)
        var position = 0
        var i = 0
        while (listConnectionSpeedMeasurementType[i] != binding.buttonConnectionSpeedList.text.toString()) {
            position = i + 1
            i++
        }

        dialog.setSingleChoiceItems(
            listConnectionSpeedMeasurementType,
            position
        ) { mDialog: DialogInterface, which: Int ->
            binding.buttonConnectionSpeedList.text = listConnectionSpeedMeasurementType[which]
            updateResult()
            mDialog.cancel()
        }.create().show()
    }

    private fun calculateTimeToDownload(size: String, speed: String, margin: Int): String? {
        var fileSize = 0.0
        var connectionSpeed = 0.0

        try {
            fileSize = size.toDouble()
            connectionSpeed = speed.toDouble()
        } catch (e: NumberFormatException) {
            e.printStackTrace()
            return null
        }

        // TODO: Move to a new object
        val fileSizeInBits: Double = fileSize * when(binding.buttonFileSizeList.text.toString()){
            getString(R.string.size_si_byte) -> SIZE_BYTE
            getString(R.string.size_si_kilo_byte) -> SIZE_SI_KILO_BYTE * SIZE_BYTE
            getString(R.string.size_si_mega_byte) -> SIZE_SI_MEGA_BYTE * SIZE_BYTE
            getString(R.string.size_si_giga_byte) -> SIZE_SI_GIGA_BYTE * SIZE_BYTE
            getString(R.string.size_si_tera_byte) -> SIZE_SI_TERA_BYTE * SIZE_BYTE
            getString(R.string.size_iec_kilo_byte) -> SIZE_IEC_KILO_BYTE * SIZE_BYTE
            getString(R.string.size_iec_mega_byte) -> SIZE_IEC_MEGA_BYTE * SIZE_BYTE
            getString(R.string.size_iec_giga_byte) -> SIZE_IEC_GIGA_BYTE * SIZE_BYTE
            getString(R.string.size_iec_tera_byte) -> SIZE_IEC_TERA_BYTE * SIZE_BYTE
            else -> { 0.0 }
        }

        // TODO: Move to a new object
        val connectionSpeedInBits: Double = connectionSpeed * when(binding.buttonConnectionSpeedList.text.toString()){
            getString(R.string.speed_bit_s) -> SPEED_BIT_S
            getString(R.string.speed_byte_s) -> SPEED_BYTE_S
            getString(R.string.speed_kilo_bit_s) -> SPEED_KILO_BIT_S
            getString(R.string.speed_mega_bit_s) -> SPEED_MEGA_BIT_S
            getString(R.string.speed_giga_bit_s) -> SPEED_GIGA_BIT_S
            getString(R.string.speed_kilo_byte_s) -> SPEED_KILO_BYTE_S
            getString(R.string.speed_mega_byte_s) -> SPEED_MEGA_BYTE_S
            getString(R.string.speed_giga_byte_s) -> SPEED_GIGA_BYTE_S
            else -> { 0.0 }
        }

        return calculateTimeToDownload(fileSizeInBits, connectionSpeedInBits, margin)
    }

    private fun calculateTimeToDownload(
        fileSize: Bits,
        connectionSpeed: Bits,
        margin: Int) : String {

        if(margin < 0 || margin > 100)
            return "Error"

        val _margin = (1.0 + (margin.toDouble() / 100))
        val timeInSeconds: Seconds = ((fileSize) / connectionSpeed) * _margin
        return timeToSting(timeInSeconds)
    }

    private fun timeToSting(timeInSeconds: Double): String {
        val minutes = (timeInSeconds / TIME_MINUTE)
        val hours = (timeInSeconds / TIME_HOUR)
        val days = (timeInSeconds / TIME_DAY)

        val day = getString(R.string.day)
        val hour = getString(R.string.hour)
        val minute = getString(R.string.minute)
        val second = getString(R.string.second)

        return if (timeInSeconds > TIME_YEAR) {
            getString(R.string.highTime)
        } else if (timeInSeconds < 1) {
            getString(R.string.lowTime)
        } else if (days >= 1) {
            floor(days).toInt()
                .toString() + "" + day + " " + floor(days * 24 % 24)
                .toInt() + "" + hour + " " + floor(days * 24 * 60 % 60)
                .toInt() + "" + minute + " " + floor(days * 24 * 60 * 60 % 60)
                .toInt() + "" + second
        } else if (hours >= 1) {
            floor(hours).toInt()
                .toString() + "" + hour + " " + floor(hours * 60 % 60)
                .toInt() + "" + minute + " " + floor(hours * 60 * 60 % 60)
                .toInt() + "" + second
        } else if (minutes >= 1) {
            floor(minutes).toInt().toString() + "" + minute + " " + floor(
                minutes * 60 % 60
            ).toInt() + "" + second
        } else if (timeInSeconds >= TIME_SECONDS) {
            floor(timeInSeconds).toInt().toString() + "" + second
        } else getString(R.string.noTime)
    }

    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}

    override fun afterTextChanged(s: Editable) {
        updateResult()
    }
    override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
        binding.errorMarginDisplay. text = getString(R.string.errorMargin, progress)
    }
    override fun onStartTrackingTouch(seekBar: SeekBar?) {}

    override fun onStopTrackingTouch(seekBar: SeekBar?) {
        updateResult()
    }

    companion object {
        private const val DEFAULT_ERROR_MARGIN = 5

        // FILE SIZE
        private const val SIZE_BIT = 1.0
        private const val SIZE_BYTE = 8.0

        // SI -> 1KB = 10^3
        private const val SIZE_SI_KILO_BYTE : Double = 1000.0 // 10^3 = 1.000
        private const val SIZE_SI_MEGA_BYTE : Double = 1000000.0 // 10^6 = 1.000.000
        private const val SIZE_SI_GIGA_BYTE : Double = 1000000000.0 // 10^9 = 1.000.000
        private const val SIZE_SI_TERA_BYTE : Double = 1000000000000.0 // 10^12 = 1.000.000

        // IEC 600274-2 -> 1KiB = 2^10
        private const val SIZE_IEC_KILO_BYTE : Double = 1024.0 // 2^10 = 1.024
        private const val SIZE_IEC_MEGA_BYTE : Double = 1048576.0 // 2^20 = 1.048.576
        private const val SIZE_IEC_GIGA_BYTE : Double = 1073741824.0 // 2^30 = 1.073.741.824
        private const val SIZE_IEC_TERA_BYTE : Double = 1099511627776.0 // 2^40 = 1.099.511.627.776

        // CONNECTION SPEED
        private const val SPEED_BIT_S = SIZE_BIT
        private const val SPEED_KILO_BIT_S = SIZE_SI_KILO_BYTE
        private const val SPEED_MEGA_BIT_S = SIZE_SI_MEGA_BYTE
        private const val SPEED_GIGA_BIT_S = SIZE_SI_GIGA_BYTE

        private const val SPEED_BYTE_S = SIZE_BYTE
        private const val SPEED_KILO_BYTE_S = SIZE_SI_KILO_BYTE * SIZE_BYTE
        private const val SPEED_MEGA_BYTE_S = SIZE_SI_MEGA_BYTE * SIZE_BYTE
        private const val SPEED_GIGA_BYTE_S = SIZE_SI_GIGA_BYTE * SIZE_BYTE

        // TIME
        private const val TIME_SECONDS = 1.0
        private const val TIME_MINUTE = TIME_SECONDS * 60
        private const val TIME_HOUR = TIME_MINUTE * 60
        private const val TIME_DAY = TIME_HOUR * 24
        private const val TIME_YEAR = TIME_DAY * 365
    }
}

typealias Bits = Double
typealias Bytes = Double
typealias Seconds = Double